rfd/0029-account-lifecycle.md

```diff
@@ -0,0 +1,231 @@
+
+Once signed up, users will be presented with the recovery tokens printed on the
+screen with a message:
+
+```
+Please save these account recovery tokens in a safe offline place.
+You can use each token once if you loose your second factor or a password.
+
+* tele-example-teleport-sh-cosponsor-happening-moocher-undrilled-royal-depravity-frustrate-starting
+* tele-example-teleport-sh-factoid-shorten-koala-explore-retinal-lazy-coauthor-brethren
+* tele-example-teleport-sh-thinness-suspect-shrink-dwindle-awning-engulf-thrift-spoiling
+```
+
+Customers can use cloud prefix in case if they forget the domain name as well.
+Teleport cloud scanners can look for `tele-` prefix and alert customers with
+domain names in the prefix about leaked tokens.
+
+* [ ] All 3 tokens are regenerated in batch every time account recovery is successful.
+* [ ] Each token can only be used one time successfully or unsuccessfully.
+* [ ] Teleport should store tokens encrypted with `crypto/bcrypt` and compare using constant-time compare.
+* [ ] Trying a wrong token 3 times in a row should lead to a temporary account lock following
+the same failed account login rate-limiting that exists for local users right now.
+* [ ] The reset endpoints should be rate-limited to prevent brute-force scans.
+* [ ] UI should make it clear that these tokens are important and if lost, the account can not be recovered.
+
+### Recovery scenarios
+
+The first two recovery scenarios work for every system user, reducing the load
+on Teleport system administrators as well.
+
+**Emails and usernames**
+
+Teleport Cloud assumes that username is a user email.
+When sending emails, Teleport cloud will send an email to the username if it contains @ sign.
+Self hosted Teleport will not send any emails.
+
+**Account owners**
+
+Teleport considers any user with delete `account` resource permission
+as account owner. This doc refers to those users as account owner users.
+
+The `account` resource is a new resource that only allows to request
+account cancellation.
+
+We will recommend to have several account owners, users who do not have system administration
+privileges and can't modify roles, or users.
+
+Any other users are referred to as simply users.
+
+**User lost a second factor**
+
+If any user lost a second factor, but not password, they can add a new second factor using
+a recovery token:
+
+* Login with a username and password.
+* Instead of pressing second factor, enter a one-time recovery token
+* Enroll new second factor.
+* All recovery tokens are no longer valid. The new set of recovery tokens are printed to the user.
+* User has to confirm that they have copied recovery tokens.
+* The session is not initiated. A user is redirected to a login form
+with UI displaying a message to login using new credentials.
+
+* If the operation is aborted after user enters the token, but fails to enter the password,
+the valid token is no longer used, but remaining tokens can be used.
```

Comment:
Above, the user enters username and password first. They don't get to consume a token before passing the password prompt

Reply:
I wonder if this leaks what is/is not a valid password? IE if I need to enter a valid password to get to the token recovery screen. On the other hand, I think that's how many other systems works anyways, which I suspect is more an artifact of 2fa being optionally enabled. Enter the password, see the account has 2fa, so then challenge for 2fa. 

Don't know if it's worth pursuing or not.

Reply:
I think it's fine for now - after all, they will be locked out after several invalid attempts, so would be hard to brute-force.